PHPBENCH: Add a benchmark test suite

sjokkateer opened the pull request:
Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description
This PR adds a small benchmark test suite to the project.

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

sjokkateer committed:
COMPOSER: Add phpbench dependency

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

sjokkateer committed:
PHPBENCH: Add configuration file

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

sjokkateer committed:
COMPOSER: Add benchmarks autoload path & benchmark command

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

sjokkateer committed:
GITIGNORE: Add .phpbench directory

Such that people can locally test for performance
regressions against different branches etc.

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

Ocramius reviewed (changes requested):
This installs the tooling, but none of the benchmarks :D 

Need at least one code path to be verified by a benchmark (ideally the code path from #64 ? )

sjokkateer commented:
> This installs the tooling, but none of the benchmarks :D
> 
> Need at least one code path to be verified by a benchmark (ideally the code path from #64 ? )

Yes :P I'll get back to the two open PRs on laminas-di soon!

Ocramius commented:
Thanks, wasn't sure if I had missed something :+1:

sjokkateer committed:
BENCHMARKS: Add benchmark class to test `ClassDefinition::getParameters()`

The testing is done before parameters are transparently cached and for three
unique conditions: class with constructor parameters, without constructor paramerters,
and a class without constructor.

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

sjokkateer commented:
@Ocramius What do you think of the added benchmark tests for testing #64 ?

Ocramius commented on README.md:
````diff
@@ -47,3 +47,15 @@ If you need these features combine it with another IoC container such as
 
 - File issues at https://github.com/laminas/laminas-di/issues
 - Documentation is at https://docs.laminas.dev/laminas-di/
+
+## Benchmarks
+
+We provide scripts for benchmarking laminas-di using the
+[PHPBench](https://github.com/phpbench/phpbench) framework; these can be
+found in the `benchmarks/` directory.
+
+To execute the benchmarks you can run the following command:
+
+```bash
+$ vendor/bin/phpbench run --report=aggregate
+```
````
I get the enthusiasm, but we do have a CI pipeline for that. If we document benchmarks and build tools, we should probably do so in a more centralized location, or `CONTRIBUTING.md`. @froschdesign may perhaps help in finding a better home for this, but `README.md` should be very minimal.

sjokkateer replied:
@Ocramius Thanks for your time and review. I'll get back to resolving the things you mentioned later today!

froschdesign replied:
@sjokkateer has done everything right so far, because he has taken the text from other packages.
But it is correct to add the text to centralized location like https://github.com/laminas/.github/blob/-/CONTRIBUTING.md

sjokkateer replied:
I removed the commit that included this in the `README.md`.

Ocramius commented on phpbench.json:
```diff
@@ -0,0 +1,7 @@
+{
```
We should probably add this to `.gitattributes` (as `export-ignore`)

Ocramius commented on benchmarks/BenchAsset/ClassDefinitionWithMultipleConstructorParameters.php:
```diff
@@ -0,0 +1,24 @@
+<?php
```
The whole `benchmarks/` directory should be in `.gitattributes` (as `export-ignore`)

sjokkateer replied:
Done

Ocramius commented on benchmarks/GetParametersClassDefinitionWithMultipleConstructorParametersBench.php:
```diff
@@ -0,0 +1,32 @@
+ * @Iterations(20)
+ * @Warmup(2)
+ */
+class GetParametersClassDefinitionWithMultipleConstructorParametersBench
+{
+    private ClassDefinition $cd;
+
+    public function __construct()
+    {
+        $this->cd = new ClassDefinition(
+            ClassDefinitionWithMultipleConstructorParameters::class
+        );
+    }
+
+    public function benchGetParametersBeforeTransparentCaching(): void
+    {
+        $cd = clone $this->cd;
+
+        $cd->getParameters();
+    }
```
Overall OK on approach, but I think we should have these `ClassDefinition#getParameters()` all in a single `*Bench` class, where each benchmarked `ClassDefinition` has a new method. 

This should keep the tests grouped together in a better way 👍

sjokkateer replied:
@Ocramius Cool, thanks for your time and review. I'll get back to resolving the things you mentioned later today!

sjokkateer replied:
@Ocramius let me know if the new [ClassDefinitionBench::class](https://github.com/laminas/laminas-di/pull/65/files#diff-9995f8946bfffe1ef3a1e2359f9f946c049cbbf5ae05ae9961c0d9350aaeeb65) is in line with your expectations.

Ocramius commented on benchmarks/GetParametersClassDefinitionWithMultipleConstructorParametersBench.php:
```diff
@@ -0,0 +1,32 @@
+use LaminasBench\Di\BenchAsset\ClassDefinitionWithMultipleConstructorParameters;
+
+/**
+ * @Revs(1000)
+ * @Iterations(20)
+ * @Warmup(2)
```
We should probably declare these ourselves, rather than here - phpbench allows setting revs, warmup, iterations, etc in the CLI. 

Let's keep this minimal here 👍

Ocramius reviewed (changes requested):
This is looking very good! Thanks @sjokkateer

My suggestions are mostly on minimal stuff to be adjusted, but after we've done that, we should have the pre/post results of #64 (and further improvements that you keep doing on this codebase)!

sjokkateer commented:
@Ocramius Cool, thanks for your time and review. I'll get back to resolving the things you mentioned later today!

sjokkateer committed:
GITATTRIBUTES: Add phpbench config & benchmark dir with export-ignore

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

sjokkateer committed:
BENCHMARKS: Move ClassDefinition related methods into one class and remove old classes

Related to the previously used GetParameters* classes.

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

Ocramius reviewed (approved):
LGTM :+1: 

:ship: 

Thanks @sjokkateer!
